### ISSUE TITLE
Create CAM for network trained on timeseries

### ISSUE THREAD

**Issue by kalfasyan** [jacobgil/pytorch-grad-cam#233]
Would it be possible to handle networks trained for timeseries classification? Example CNN below:
```
class Cnn1d(nn.Module):

    def __init__(self, outputs=2):
        super(Cnn1d, self).__init__()
        self.outputs = outputs
        self.conv1 = nn.Conv1d(1, 16, 3)
        self.bn1 = nn.BatchNorm1d(16)
        self.pool1 = nn.MaxPool1d(2)

        self.conv2 = nn.Conv1d(16, 32, 3)
        self.bn2 = nn.BatchNorm1d(32)
        self.pool2 = nn.MaxPool1d(2)

        self.conv3 = nn.Conv1d(32, 64, 3)
        self.bn3 = nn.BatchNorm1d(64)
        self.pool3 = nn.MaxPool1d(2)

        self.conv4 = nn.Conv1d(64, 128, 3)
        self.bn4 = nn.BatchNorm1d(128)
        self.pool4 = nn.MaxPool1d(2) 

        self.dropout = nn.Dropout()
        self.avgPool = nn.AvgPool1d(127)
        self.fc1 = nn.Linear(256, self.outputs)

    def forward(self, x):

        x = self.conv1(x)
        x = F.relu(self.bn1(x))
        x = self.pool1(x)

        x = self.conv2(x)
        x = F.relu(self.bn2(x))
        x = self.pool2(x)

        x = self.conv3(x)
        x = F.relu(self.bn3(x))
        x = self.pool3(x)

        x = self.conv4(x)
        x = F.relu(self.bn4(x))
        x = self.pool4(x)

        x = self.dropout(x)
        x = self.avgPool(x)
        x = x.view(x.shape[0], -1)
        x = self.fc1(x)

        return x
```

**Comment by jacobgil** [comment]
Hi,
Interesting question.

I was about to say that no, since this package is aimed for images,

But actually we can treat the timeseries as an image with one row.
Then we can "resize" that image to see stripes, where a hot stripe means that element was important.

The trick would be to write a custom reshape_transform, that is responsible for getting the tensors from the
target layer, and transforming them into a 2D image.
The target layer would be the dropout layer, since that's what used just before the global average pooling in this case.

I think something like this should work:
```python
def reshape_transform(tensor):
    image_with_single_row = tensor[:, None, :, :]
    # Lets make the time series into an image with 16 rows for easier visualization on screen later
    target_size = 16, tensor.size(1)
    return torch.nn.functional.interpolate(image_with_single_row, target_size, mode='bilinear')
```

**Comment by jacobgil** [comment]
Btw, it would be awesome to have this example as a notebook example.

Let me know if I can help

**Comment by kalfasyan** [comment]
Thanks for the reply :) So this is what I've got so far, can you let me know if it looks logical to you?
```
from pytorch_grad_cam import GradCAM, ScoreCAM, GradCAMPlusPlus, AblationCAM, XGradCAM, EigenCAM, FullGrad
from pytorch_grad_cam.utils.model_targets import ClassifierOutputTarget
from pytorch_grad_cam.utils.image import show_cam_on_image
from torchvision.models import resnet50

n = np.random.randint(0,len(train_dataset))

target_layers = [model.bn8] # using a bit deeper model in this example

input_tensor = train_dataset[n][0].unsqueeze(0) # torch.Size([1, 1, 5000])
targets = None

cam = GradCAM(model=model, target_layers=target_layers, use_cuda=True, reshape_transform=reshape_transform) 
# using your reshape function from above @jacobgil 

grayscale_cam = cam(input_tensor=input_tensor, targets=targets)

plt.figure()
plt.plot(scale(input_tensor.squeeze().numpy()))
plt.plot(grayscale_cam.T.squeeze())
```
Btw, I have binary classification and the model has a single output. My signal is the wingbeat recording of a fly using an optical sensor. A single recording is 5000 measurements from a sensor and it can have flight and noise. In this example below, the flight is up to sample 3000 (x-axis) and then you have noise. I scale my signal (blue) from 0 to 1 to overlay the gradcam output (orange). It seems that I should multiply by -1 to get the weights to fall in the interesting signal region and not the noise.
![gradcamexample](https://user-images.githubusercontent.com/5652952/163689652-1d3e6dab-b7ea-41ea-ade3-df881a46ce86.png)

**Comment by jacobgil** [comment]
Oh ok, so the samples here are 1D features.
I think what grad-cam is doing here, is just taking the gradient.
So it's not really grad-cam, it collapses into the older "gradient of output vs input" saliency method.

And apparently the gradients are higher in the end of the time series - does that make sense in terms of what you expect from the model?

**Comment by kalfasyan** [comment]
So this is interpreted more like "this is where the model is more wrong and needs to change drastically" instead of "this is what contributes more to its final prediction", am I right to read it like this?
